server/service/osquery.go

```diff
@@ -523,58 +532,84 @@ func (svc *Service) GetDistributedQueries(ctx context.Context) (map[string]strin
 
 	policyQueries, err := svc.policyQueriesForHost(ctx, host)
 	if err != nil {
-		return nil, 0, osqueryError{message: err.Error()}
+		return nil, nil, 0, osqueryError{message: err.Error()}
 	}
 	for name, query := range policyQueries {
 		queries[hostPolicyQueryPrefix+name] = query
 	}
 
-	accelerate := uint(0)
+	accelerate = uint(0)
 	if host.Hostname == "" || host.Platform == "" {
 		// Assume this host is just enrolling, and accelerate checkins
 		// (to allow for platform restricted labels to run quickly
 		// after platform is retrieved from details)
 		accelerate = 10
 	}
 
-	return queries, accelerate, nil
+	// The way osquery's distributed "discovery" queries work is:
+	// If len(discovery) > 0, then only those queries that have a "discovery"
+	// query and return more than one row are executed on the host.
+	//
+	// Thus, we set the alwaysTrueQuery for all queries, except for those where we set
+	// an explicit discovery query (e.g. orbit_info, google_chrome_profiles).
+	for name := range queries {
+		discoveryQuery := discovery[name]
+		if discoveryQuery == "" {
+			discoveryQuery = alwaysTrueQuery
+		}
+		discovery[name] = discoveryQuery
+	}
+
+	return queries, discovery, accelerate, nil
 }
 
+const alwaysTrueQuery = "SELECT 1"
+
 // detailQueriesForHost returns the map of detail+additional queries that should be executed by
 // osqueryd to fill in the host details.
-func (svc *Service) detailQueriesForHost(ctx context.Context, host *fleet.Host) (map[string]string, error) {
+func (svc *Service) detailQueriesForHost(ctx context.Context, host *fleet.Host) (queries map[string]string, discovery map[string]string, err error) {
 	if !svc.shouldUpdate(host.DetailUpdatedAt, svc.config.Osquery.DetailUpdateInterval, host.ID) && !host.RefetchRequested {
-		return nil, nil
+		return nil, nil, nil
 	}
 
 	config, err := svc.ds.AppConfig(ctx)
 	if err != nil {
-		return nil, ctxerr.Wrap(ctx, err, "read app config")
+		return nil, nil, ctxerr.Wrap(ctx, err, "read app config")
 	}
 
-	queries := make(map[string]string)
+	queries = make(map[string]string)
+	discovery = make(map[string]string)
+
 	detailQueries := osquery_utils.GetDetailQueries(config, svc.config)
 	for name, query := range detailQueries {
 		if query.RunsForPlatform(host.Platform) {
-			queries[hostDetailQueryPrefix+name] = query.Query
+			queryName := hostDetailQueryPrefix + name
+			queries[queryName] = query.Query
+			discoveryQuery := query.Discovery
+			if discoveryQuery == "" {
+				discoveryQuery = alwaysTrueQuery
```

Comment:
Random thought: I wonder if we could use discovery queries instead of per platform logic in here and let osquery respond to whatever makes sense.